### DISCUSSION TITLE
Is HighFive thread-safe?

### DISCUSSION THREAD

**Discussion by YoniSp** [BlueBrain/HighFive#614] [Questions] [answered]
Hi,

When I trying to open/using the same file on 2 threads using H5Easy api
the API commands which has been seen to be problematic are:
`H5Easy::File` and `H5Easy::load`

Error mesage is:
```
HDF5-DIAG: Error detected in HDF5 (1.13.2) thread 0:
  #000: /root/Downloads/hdf5-hdf5-1_13_2/src/H5F.c line 837 in H5Fopen(): unable to synchronously open file
    major: File accessibility
    minor: Unable to open file
  #001: /root/Downloads/hdf5-hdf5-1_13_2/src/H5F.c line 797 in H5F__open_api_common(): unable to open file
    major: File accessibility
    minor: Unable to open file
  #002: /root/Downloads/hdf5-hdf5-1_13_2/src/H5VLcallback.c line 3686 in H5VL_file_open(): open failed
    major: Virtual Object Layer
    minor: Can't open object
  #003: /root/Downloads/hdf5-hdf5-1_13_2/src/H5VLcallback.c line 3498 in H5VL__file_open(): open failed
    major: Virtual Object Layer
    minor: Can't open object
  #004: /root/Downloads/hdf5-hdf5-1_13_2/src/H5VLnative_file.c line 128 in H5VL__native_file_open(): unable to open file
    major: File accessibility
    minor: Unable to open file
  #005: /root/Downloads/hdf5-hdf5-1_13_2/src/H5Fint.c line 1992 in H5F_open(): file close degree doesn't match
    major: File accessibility
    minor: Unable to initialize object

Process finished with exit code 1
```

HDF5 version: hdf5-1_13_2

My question is : does `H5Easy` natively supports multi threading.
If yes how can I overcome the issue.
If no how can I get H5Easy  multi threading support or how to use HighFive when using multi threading / multi process?

**Comment by 1uc** [comment, marked as answer]
There's two options:
* ~use a thread safe build of HDF5,~
* implement your own API lock.

Personally, since HDF5 is only thread-safe, but without any concurrency, I prefer to implement the locking myself. This way I understand the locking strategy; and the STL supports easy to use mutexes.

You can look into:
https://docs.hdfgroup.org/hdf5/v1_12/_m_t.html

As a side note: when opening there could be a race, if you're opening and truncating the files.

**Comment by YoniSp** [comment]
Hi I compiled HDF5 with  thread safe but then there is no c++ support
When I compiled without it 
I got a new error:
terminate called after throwing an instance of 'HighFive::DataSetException'
  what():  Unable to open the dataset "/data_10/a": (Virtual Object Layer) Bad value

what does it means?

**Comment by 1uc** [comment]
I'm sorry, it's very hard to know what you're attempting to do. Please provide a minimal example, including build instructions, etc. Please describe carefully what you expect to happen and what you've tried. I find the process of writing a precise question often answers the question itself.
